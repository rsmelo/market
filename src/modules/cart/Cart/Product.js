import React from 'react'
import PropTypes from 'prop-types'

import Button from '../../../components/Button'
import style from './Product.style.css'

const Product = ({
  id,
  image,
  name,
  price,
  seller,
  onRemove,
}) => (
  <tr>
    <td>
      <img
        src={image}
        alt={name}
        className={style.image}
      />
    </td>
    <td className={style.productCell}>{name}</td>
    <td className={style.productCell}>R$ {price}</td>
    <td className={style.productCell}>
      <Button onClick={() => onRemove(id, seller.id)}>
        Remover do carrinho
      </Button>
    </td>
  </tr>
)

Product.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  image: PropTypes.string.isRequired,
  seller: PropTypes.shape({
    id: PropTypes.string.isRequired,
  }).isRequired,
  onRemove: PropTypes.func.isRequired,
}
export default Product
