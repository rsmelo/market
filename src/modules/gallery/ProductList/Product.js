import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

import style from './Product.style.css'

const Product = ({
  id,
  name,
  price,
  image,
}) => (
  <li className={style.product}>
    <Link to={`/product/${id}`}>
      <img
        className={style.image}
        src={image}
        alt={name}
      />
    </Link>
    <div className={style.information}>
      <Link to={`/product/${id}`}>
        <h2 className={style.name}>{name}</h2>
        <span className={style.price}>R$ {price}</span>
      </Link>
    </div>
  </li>
)

Product.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  image: PropTypes.string.isRequired,
}

export default Product
