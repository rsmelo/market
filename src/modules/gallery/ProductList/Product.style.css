.name {
  position: relative;
  overflow: hidden;
  white-space: nowrap;
  display: block;
  font-size: 18px;
  line-height: 1.2;
  color: #7d7a77;
  margin: 0 0 7px;
}

.price {
  font-size: 16px;
  color: #f77274;
  font-weight: bold;
}

.image {
  position: relative;
  display: block;
  width: 276px;
  height: 294px;
  background: #f4f2f0;
  border-radius: 3px;
}
