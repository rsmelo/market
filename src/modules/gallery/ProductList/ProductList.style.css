.productList {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  max-width: 75%;
  padding: 1em 2em;
}
