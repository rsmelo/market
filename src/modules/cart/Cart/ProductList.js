import React from 'react'
import PropTypes from 'prop-types'

import Product from './Product'
import style from './ProductList.style.css'
import calculateTotal from '../../../utils/calculateTotal'


const ProductList = ({ products, removeProduct }) => (
  <table className={style.productList}>
    <thead className={style.header}>
      <tr>
        <th colSpan="2">Produto(s)</th>
        <th >Pre√ßo</th>
        <th />
      </tr>
    </thead>
    <tbody>
      {products.map(product => (
        <Product
          key={product.id}
          {...product}
          onRemove={removeProduct}
        />
      ))
      }
    </tbody>
    <tfoot className={style.footer}>
      <tr>
        <td colSpan="2" />
        <td />
        <td className={style.totalLabel}>
          <span>Total</span> R$ {calculateTotal(products)}
        </td>
      </tr>
    </tfoot>
  </table>
)

ProductList.propTypes = {
  products: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  removeProduct: PropTypes.func.isRequired,
}

export default ProductList
