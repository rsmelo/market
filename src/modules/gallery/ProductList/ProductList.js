import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import Product from './Product'

class ProductList extends PureComponent {
  constructor (props) {
    super(props)
    this.renderProducts = this.renderProducts.bind(this)
  }

  componentDidMount () {
    this.props.getProducts()
  }

  renderProducts () {
    const { products } = this.props
    return (products && products.map(product => (
      <Product {...product} key={product.id} />
    )))
  }

  render () {
    return (
      <ul>
        {this.renderProducts()}
      </ul>
    )
  }
}

ProductList.defaultProps = {
  products: null,
}

ProductList.propTypes = {
  products: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
  })),
  getProducts: PropTypes.func.isRequired,
}

export default ProductList
