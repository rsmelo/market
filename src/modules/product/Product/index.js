import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'

import Product from './Product'

import {
  getGalleryProductById,
  getGalleryIsFetching,
  getExistProductInCart,
} from '../../combinedReducers'
import { getProduct } from '../actions'
import { addProduct, removeProduct } from '../../cart/actions'

const mapStateToProps = (state, ownProps) => {
  const product = getGalleryProductById(state, ownProps.match.params.id)
  return {
    product,
    isFetching: getGalleryIsFetching(state),
    addedToCart: product ?
      getExistProductInCart(state, product.id, product.seller.id) :
      false,
  }
}

const mapDispatchToProps = {
  getProduct,
  addProduct,
  removeProduct,
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Product))
