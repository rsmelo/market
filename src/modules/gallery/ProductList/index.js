import { connect } from 'react-redux'

import ProductList from './ProductList'

import { getProducts } from '../actions'
import { getGalleryProducts } from '../../combinedReducers'

const mapStateToProps = state => ({
  products: getGalleryProducts(state),
})

const mapDispatchToProps = {
  getProducts,
}


export default connect(mapStateToProps, mapDispatchToProps)(ProductList)
