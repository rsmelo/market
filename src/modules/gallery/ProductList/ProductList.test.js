import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

import ProductList from './ProductList'
import Product from './Product'

jest.mock('react-router-dom', () => ({ Link: 'Link' }))

describe('ProductList', () => {
  it('should render correctly', () => {
    const getProducts = jest.fn()
    const products = [
      {
        id: 'a12090',
        name: 'a product',
        price: 10,
        image: 'image-url',
        description: 'test',
      },
      {
        id: 'a120902',
        name: 'a product2',
        price: 10,
        image: 'image-url2',
        description: 'test',
      },
    ]
    const props = {
      products,
      getProducts,
      isFetching: true,
    }

    const tree = renderer
      .create(<ProductList {...props} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('should call getProducts when mount', () => {
    const getProducts = jest.fn()
    shallow(<ProductList getProducts={getProducts} isFetching={false} />)
    expect(getProducts).toHaveBeenCalled()
  })

  it('should a Product to each item in products property', () => {
    const getProducts = jest.fn()
    const products = [
      {
        id: 'a12090',
        name: 'a product',
        price: 10,
        image: 'image-url',
        description: 'test',
      },
      {
        id: 'a120902',
        name: 'a product2',
        price: 10,
        image: 'image-url2',
        description: 'test',
      },
    ]

    const props = {
      getProducts,
      products,
      isFetching: false,
    }

    const component = shallow(<ProductList {...props} />)
    const productsComponents = component.find(Product)
    expect(productsComponents).toHaveLength(2)
  })
})
