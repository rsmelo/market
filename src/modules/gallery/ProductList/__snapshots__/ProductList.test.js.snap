// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList should render correctly 1`] = `
<ul
  className={undefined}
>
  <li
    className={undefined}
  >
    <Link
      to="/product/a12090"
    >
      <img
        alt="a product"
        className={undefined}
        src="image-url"
      />
    </Link>
    <div
      className={undefined}
    >
      <Link
        to="/product/a12090"
      >
        <h2
          className={undefined}
        >
          a product
        </h2>
        <span
          className={undefined}
        >
          R$ 
          10
        </span>
      </Link>
    </div>
  </li>
  <li
    className={undefined}
  >
    <Link
      to="/product/a120902"
    >
      <img
        alt="a product2"
        className={undefined}
        src="image-url2"
      />
    </Link>
    <div
      className={undefined}
    >
      <Link
        to="/product/a120902"
      >
        <h2
          className={undefined}
        >
          a product2
        </h2>
        <span
          className={undefined}
        >
          R$ 
          10
        </span>
      </Link>
    </div>
  </li>
</ul>
`;
